Mathlib/NumberTheory/ModularForms/DedekindEta.lean

```diff
@@ -0,0 +1,143 @@
+/-
+Copyright (c) 2024 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck, David Loeffler
+-/
+
+import Mathlib.Algebra.Order.Ring.Star
+import Mathlib.Analysis.CStarAlgebra.Classes
+import Mathlib.Analysis.Complex.LocallyUniformLimit
+import Mathlib.Analysis.Complex.UpperHalfPlane.Exp
+import Mathlib.Analysis.NormedSpace.MultipliableUniformlyOn
+import Mathlib.Data.Complex.FiniteDimensional
+
+/-!
+# Dedekind eta function
+
+## Main definitions
+
+* We define the Dedekind eta function as the infinite product
+`η(z) = q ^ 1/24 * ∏' (1 - q ^ (n + 1))` where `q = e ^ (2πiz)` and `z` is in the upper half-plane.
+The product is taken over all non-negative integers `n`. We then show it is non-vanishing and
+differentiable on the upper half-plane.
+
+## References
+* [F. Diamond and J. Shurman, *A First Course in Modular Forms*][diamondshurman2005]
+-/
+
+open UpperHalfPlane TopologicalSpace Set MeasureTheory intervalIntegral
+ Metric Filter Function Complex
+
+open scoped Interval Real NNReal ENNReal Topology BigOperators Nat
+
+local notation "𝕢" => Periodic.qParam
+
+local notation "ℍₒ" => complexUpperHalfPlane
+
+/-- The q term inside the product defining the eta function. It is defined as
+`eta_q n z = e ^ (2 π i (n + 1) z)`. -/
+noncomputable abbrev eta_q (n : ℕ) (z : ℂ) := (𝕢 1 z) ^ (n + 1)
```

Comment:
This should be in some namespace (maybe `ModularForm.eta_q`) to avoid cluttering the global namespace (& ditto `etaProdTerm`)